src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -355,24 +79,13 @@ class TypeComparer(initctx: Context) extends DotClass {
     if ((tp2 eq tp1) ||
         (tp2 eq WildcardType) ||
         (tp2 eq AnyType) && tp1.isValueType) return true
-    isSubType(tp1, tp2)
-  }
-
-  def isSubTypeWhenFrozen(tp1: Type, tp2: Type): Boolean = {
-    val saved = frozenConstraint
-    frozenConstraint = true
     try isSubType(tp1, tp2)
-    finally frozenConstraint = saved
+    finally
+      if (Config.checkConstraintsSatisfiable)
+        assert(isSatisfiable, constraint.show)
   }
 
-  def isNonBottomSubType(tp1: Type, tp2: Type): Boolean =
-    !(tp2 isRef NothingClass) && isSubType(tp1, tp2)
-
-  private def traceInfo(tp1: Type, tp2: Type) =
-    s"${tp1.show} <:< ${tp2.show}" +
-    (if (ctx.settings.verbose.value) s" ${tp1.getClass} ${tp2.getClass}${if (frozenConstraint) " frozen" else ""}" else "")
-
-  def isSubType(tp1: Type, tp2: Type): Boolean = /*>|>*/ ctx.traceIndented(s"isSubType ${traceInfo(tp1, tp2)}", subtyping) /*<|<*/ {
+  def isSubType(tp1: Type, tp2: Type): Boolean = ctx.traceIndented(s"isSubType ${traceInfo(tp1, tp2)} ${if (Config.verboseExplainSubtype) s" ${tp1.getClass}, ${tp2.getClass}" else ""}", subtyping) /*<|<*/ {
```

Comment:
Should `isSubType` be `protected`? Only `topLevelSubType` is used outside the class and its parent traits.